src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -1623,89 +1469,184 @@ private static string DeterminePropNameIdentifier(string runtimePropName)
                 return sb.ToString();
             }
 
-            private JsonPrimitiveTypeKind? GetPrimitiveTypeKind(Type type)
+            private JsonPrimitiveTypeKind? GetPrimitiveTypeKind(ITypeSymbol type)
             {
-                if (_numberTypes.Contains(type))
+                if (type.IsNumberType())
+                {
                     return JsonPrimitiveTypeKind.Number;
+                }
 
-                if (type == _stringType || type == _dateTimeType || type == _dateTimeOffsetType || type == _guidType)
+                if (type.SpecialType is SpecialType.System_Boolean)
                 {
-                    return JsonPrimitiveTypeKind.String;
+                    return JsonPrimitiveTypeKind.Boolean;
                 }
 
-                if (type == _booleanType)
+                if (type.SpecialType is SpecialType.System_Char)
                 {
-                    return JsonPrimitiveTypeKind.Boolean;
+                    return JsonPrimitiveTypeKind.Char;
                 }
 
-                if (type == _byteArrayType)
+                SymbolEqualityComparer cmp = SymbolEqualityComparer.Default;
+
+                if (type.SpecialType is SpecialType.System_String or SpecialType.System_DateTime ||
+                     cmp.Equals(type, _knownSymbols.DateTimeOffsetType) || cmp.Equals(type, _knownSymbols.GuidType))
                 {
-                    return JsonPrimitiveTypeKind.ByteArray;
+                    return JsonPrimitiveTypeKind.String;
                 }
 
-                if (type == _charType)
+                if (cmp.Equals(type, _knownSymbols.ByteArrayType))
                 {
-                    return JsonPrimitiveTypeKind.Char;
+                    return JsonPrimitiveTypeKind.ByteArray;
                 }
 
                 return null;
             }
 
-            private void PopulateNumberTypes()
+            private static string GetTypeInfoPropertyName(ITypeSymbol type)
+            {
+                if (type is IArrayTypeSymbol arrayType)
+                {
+                    int rank = arrayType.Rank;
+                    string suffix = rank == 1 ? "Array" : $"Array{rank}D"; // Array, Array2D, Array3D, ...
+                    return GetTypeInfoPropertyName(arrayType.ElementType) + suffix;
+                }
+
+                if (type is not INamedTypeSymbol namedType || !namedType.IsGenericType)
+                {
+                    return type.Name;
+                }
+
+                StringBuilder sb = new();
+
+                string name = namedType.Name;
+
+                sb.Append(name);
+
+                foreach (ITypeSymbol genericArg in namedType.GetAllTypeArgumentsInScope())
+                {
+                    sb.Append(GetTypeInfoPropertyName(genericArg));
+                }
+
+                return sb.ToString();
+            }
+
+            private static bool TryGetDeserializationConstructor(
```

Comment:
Why not maintain an abstraction for both symbols and `Type`s, similar to `TraverseGraphWithTopologicalSort`?

Reply:
The sorting function only needs a graph abstraction to function, so it's fairly straightforward to generalize. This method on the other hand depends on many things, including querying for constructor methods, accessibility modifiers and arities. It is introducing a bit of duplication but that's probably fine (most other logic in the parser is already duplicating logic from the reflection serializer).

Reply:
Well I guess the duplication is fine if there's not much further churn.